activerecord/lib/active_record/connection_adapters/abstract/quoting.rb

```diff
@@ -157,6 +157,14 @@ def prepare_binds_for_database(binds) # :nodoc:
 
       private
 
+      def type_casted_binds(binds)
+        if binds.first.is_a?(Array)
+          binds.map { |column, value| type_cast(value, column) }
+        else
+          binds.map { |attr| type_cast(attr.value_for_database) }
```

Comment:
Since 5.0, `binds` that generated in the internal is an array of `QueryAttribute`.
This fix only addresses passing legacy `binds` explictly case (e.g. `conn.select_all("...", nil, [[column, value], ...])`) and it is sufficient fix I think.